### PR TITLE
add TKEK functions

### PR DESCRIPTION
**Pull request by adamcarlton**
add function calls for creating/getting TKEK and extend create to allow for a secure import request with TKEK

### CODE REVIEW

File: redstone/client.py [hunk 1]
```diff
@@ -418,6 +418,14 @@ def create(self, name, payload=None, raw_payload=None, root=False):
         elif payload is not None:
             data['resources'][0]['payload'] = base64.b64encode(payload).decode('utf-8')
 
+        # TKEK related fields
+        if encryption is not None:
+            data['resources'][0]['encryptionAlgorithm'] = 'RSAES_OAEP_SHA_256'
```
Comment (mrodden) [inline redstone/client.py]: No point in adding a keyword arg if you are just going to set this to a hardcoded value.

Put this set into the if block where you check if they passed a signed_nonce or iv

File: redstone/client.py [hunk 2]
```diff
@@ -397,7 +397,7 @@ def get(self, key_id):
 
         return resp.json().get('resources')[0]
 
-    def create(self, name, payload=None, raw_payload=None, root=False):
+    def create(self, name, payload=None, raw_payload=None, root=False, encryption=None, signedNonce=None, iv=None):
```
Comment (mrodden) [inline redstone/client.py]: use `signed_nonce` instead of capCase

File: redstone/client.py [hunk 3]
```diff
@@ -418,6 +418,14 @@ def create(self, name, payload=None, raw_payload=None, root=False):
         elif payload is not None:
             data['resources'][0]['payload'] = base64.b64encode(payload).decode('utf-8')
 
+        # TKEK related fields
+        if encryption is not None:
+            data['resources'][0]['encryptionAlgorithm'] = 'RSAES_OAEP_SHA_256'
+        if signedNonce is not None:
+            data['resources'][0]['signedNonce'] = signedNonce
+        if iv is not None:
+            data['resources'][0]['iv'] = iv
```
Comment (mrodden) [inline redstone/client.py]: I don't understand what these fields are... the docs indicate that there should be an `import_key` field...

https://cloud.ibm.com/docs/services/key-protect?topic=key-protect-import-root-keys#import-root-key

Also, if you need both you should have them pass a tuple of the (nonce, iv) pair or check both and raise a ValueError("'signed_nonce' and 'iv' must both be specified for importing client encrypted payloads")

File: redstone/client.py [hunk 4]
```diff
@@ -460,6 +468,21 @@ def unwrap(self, key_id, ciphertext, aad=None):
         resp = self._action(key_id, "unwrap", data)
         return base64.b64decode(resp['plaintext'].encode())
 
+    def createTKEK(self, expiration, maxAllowedRetrievals):
```
Comment (mrodden) [inline redstone/client.py]: createTKEK -> create_transport_key or maybe create_import_token

also maxAllowedRetrievals should be max_allowed_retrievals, maybe shorten it to `max_retrievals` for convenience

File: redstone/client.py [hunk 5]
```diff
@@ -460,6 +468,21 @@ def unwrap(self, key_id, ciphertext, aad=None):
...
+        data = {'expiration': expiration,
+                'maxAllowedRetrievals': maxAllowedRetrievals}
+
+        resp = self.session.post("%s/api/v2/import_token" % self.endpoint_url,
+                                 json=data)
+
+        self._validate_resp(resp)
+        return
+
+    def getTKEK(self):
```
Comment (mrodden) [inline redstone/client.py]: same as above, getTKEK -> get_transport_key/get_import_token

also, we don't need an ID here? https://cloud.ibm.com/apidocs/key-protect#retrieve-a-transport-key-by-id

**Review by mrodden — CHANGES_REQUESTED** [review]
So it looks like you are implementing an undeployed/unpublished API here... please point me at the docs and it would be great to know if I can test this somewhere.

The PR also needs some changes detailed in the inline comments.

File: redstone/client.py (hunk 3 above)
Comment (adamcarlton) [reply redstone/client.py]: we no longer use `importToken`, nonce is what's taken its place. As to nonce/iv, noted, will fix

File: redstone/client.py (hunk 5 above)
Comment (adamcarlton) [reply redstone/client.py]: we don't. we've moved to using the bluemix instance id as the key to the import_token table in the DB.

File: redstone/client.py [hunk 6]
```diff
@@ -460,6 +468,21 @@ def unwrap(self, key_id, ciphertext, aad=None):
...
 
+    def create_import_token(self, expiration, max_retrievals):
+        data = {'expiration': expiration,
+                'maxAllowedRetrievals': max_retrievals}
+
+        resp = self.session.post("%s/api/v2/import_token" % self.endpoint_url,
+                                 json=data)
+
+        self._validate_resp(resp)
+        return
```
Comment (mrodden) [inline redstone/client.py]: You should return the response here, but I don't have API references to work from so...

File: redstone/client.py [hunk 7]
```diff
@@ -418,6 +418,14 @@ def create(self, name, payload=None, raw_payload=None, root=False):
         elif payload is not None:
             data['resources'][0]['payload'] = base64.b64encode(payload).decode('utf-8')
 
+        # TKEK related fields
+        if signed_nonce is not None and iv is not None:
+            data['resources'][0]['signedNonce'] = signed_nonce
+            data['resources'][0]['iv'] = iv
+            data['resources'][0]['encryptionAlgorithm'] = 'RSAES_OAEP_SHA_256'
+        elif iv is not None and signed_nonce is None or iv is None and signed_nonce is not None:
+            ValueError("`signed_nonce` and `iv` both need to be specified for importing client encrypted payloads")
```
Comment (mrodden) [inline redstone/client.py]: Use single quotes (') not back ticks for the quoting inline... pretty sure backticks mean something different

File: redstone/client.py [hunk 8]
```diff
@@ -418,6 +418,14 @@ def create(self, name, payload=None, raw_payload=None, root=False):
         elif payload is not None:
             data['resources'][0]['payload'] = base64.b64encode(payload).decode('utf-8')
 
+        # TKEK related fields
+        if signed_nonce is not None and iv is not None:
```
Comment (mrodden) [inline redstone/client.py]: try this
```
if any([signed_nonce, iv]):
    if all([signed_nonce, iv]):
          # we have all the parameters we need
          data[....] = ...
    else:
          raise ValueError(...)
```

**Review by mrodden — CHANGES_REQUESTED** [review]

**Review by mrodden — COMMENTED** [review]
Ok. Looks reasonable. Do you need this merged for some timeline?

**Comment by adamcarlton** [comment]
not anything immediate. This needed to be added to allow for modification of `awacs` which will be what's used to test against what we have deployed.

**Pull request closed without merging**